Add multi-user support

stefanv opened the pull request:
This PR also improves the WebSocket setup.

To test this branch, you need to have https://github.com/python-social-auth/social-storage-peewee/pull/1 installed

TODO:

- [x] Move token to configuration file
- [x] Disable username (or use login/pass) for testing purposes

stefanv committed:
WIP: multi-user

stefanv committed:
Move public -> static

stefanv committed:
Complete Social Auth

stefanv committed:
Fix websocket bus

stefanv committed:
Update messaging according to Min's recommendations

- The message bus now uses PUSH (`Flow` side) and PULL (bus side) to relay
  mesages; this is more reliable than PUB/SUB used before.
- PUB/SUB is still used to send messages from the bus to websockets, but each
  server now only subscribes to messages for its connected users (in the
  unlikely but exciting event that our user base grows to thousands of
  simultaneously active users, messages will be filtered more effective in the
  ZMQ layer, rather than in Python).

stefanv committed:
Working Social Auth, along with https://github.com/python-social-auth/social-storage-peewee/pull/1

stefanv committed:
Add user profile

pep8speaks commented:
Hello @stefanv! Thanks for updating the PR.

 - In the file [`cesium_app/app_server.py`](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/app_server.py), following are the PEP8 issues :

> [Line 32:58](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/app_server.py#L32): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 33:32](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/app_server.py#L33): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 33:63](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/app_server.py#L33): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator









 - In the file [`cesium_app/handlers/dataset.py`](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/handlers/dataset.py), following are the PEP8 issues :

> [Line 52:81](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/handlers/dataset.py#L52): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 80 characters)



 - In the file [`cesium_app/handlers/mainpage.py`](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/handlers/mainpage.py), following are the PEP8 issues :

> [Line 4:1](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/handlers/mainpage.py#L4): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1









 - In the file [`cesium_app/handlers/project.py`](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/handlers/project.py), following are the PEP8 issues :

> [Line 32:25](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/handlers/project.py#L32): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 33:25](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/handlers/project.py#L33): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent



 - In the file [`cesium_app/model_util.py`](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/model_util.py), following are the PEP8 issues :

> [Line 24:17](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/model_util.py#L24): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 25:17](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/model_util.py#L25): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 25:17](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/model_util.py#L25): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 97:81](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/model_util.py#L97): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 101:81](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/model_util.py#L101): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 80 characters)
> [Line 102:31](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/model_util.py#L102): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 106:26](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/model_util.py#L106): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 107:26](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/model_util.py#L107): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 110:81](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/model_util.py#L110): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)





 - In the file [`cesium_app/tests/conftest.py`](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/tests/conftest.py), following are the PEP8 issues :

> [Line 16:23](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/tests/conftest.py#L16): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 17:23](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/tests/conftest.py#L17): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 19:1](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/tests/conftest.py#L19): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 76:81](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/tests/conftest.py#L76): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 80 characters)

 - In the file [`cesium_app/tests/fixtures.py`](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/tests/fixtures.py), following are the PEP8 issues :

> [Line 143:81](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/tests/fixtures.py#L143): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 80 characters)
> [Line 180:81](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/cesium_app/tests/fixtures.py#L180): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 80 characters)





 - In the file [`services/app.py`](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/services/app.py), following are the PEP8 issues :

> [Line 5:1](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/services/app.py#L5): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 6:1](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/services/app.py#L6): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file











 - In the file [`tools/test_frontend.py`](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/tools/test_frontend.py), following are the PEP8 issues :

> [Line 8:1](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/tools/test_frontend.py#L8): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 18:1](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/tools/test_frontend.py#L18): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 101:81](https://github.com/stefanv/cesium_web/blob/0c1ac2c53952cdcb4030105af43ea0eb06ad4cd2/tools/test_frontend.py#L101): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 80 characters)





##### Comment last updated on April 26, 2017 at 22:34 Hours UTC

stefanv committed:
Allow both single and multi-user operation

This is a first step in getting the test suite to run again

acrellin commented on cesium_app/handlers/base.py:
```diff
@@ -48,19 +46,49 @@ def prepare(self):
+                models.User.get(username=username)
+            except models.User.DoesNotExist:
+                models.User.create(username=username, email=username)
+            return 'testuser@gmail.com'
```
`return username`? doubly hard-coded

acrellin commented on static/scripts/actions.jsx:
```diff
@@ -571,7 +616,9 @@ export function fetchPredictions() {
 
-      fetch('/predictions')
+      fetch('/predictions', {
+        credentials: 'same-origin'
```
Perhaps this should be the default param, which would clean these calls up a bit.

acrellin commented:
Looking great so far, @stefanv! Nice work

acrellin committed:
Add server URL to test yaml

acrellin commented:
@stefanv that little addition gets all tests passing except one on my machine

stefanv commented:
Woah, neat!  Thanks @acrellin

stefanv committed:
Create artificial user when testing

stefanv committed:
Update testing instructions for devs

stefanv committed:
Fix JS error when there are no models

stefanv committed:
Fix favicon

stefanv committed:
Move oauth token to configuration

stefanv committed:
Add logout endpoint

stefanv committed:
Move instructions for getting OAuth credentials to config

acrellin committed:
Install newer nginx on travis-ci

Versions >= 1.7 supports `if` syntax in logging, allowing us to ignore
certain status codes in the error logs.

stefanv committed:
Remove file accidentally added back in

stefanv committed:
Define test username once

stefanv committed:
Raise error if socket auth attempted without username

stefanv committed:
Refactor models to split out utils and Peewee models/storage

stefanv committed:
Add fake oauth2 and improve supervisord interface

stefanv committed:
Improve login box design

stefanv committed:
Generate random auth code in fake oauth2

stefanv committed:
Do not create dummy user as fixture in test suite; add users via Python Social Auth only!

acrellin committed:
Click login after webdriver initialization

stefanv commented on cesium_app/tests/conftest.py:
```diff
@@ -35,11 +38,23 @@ def close():
+        # Already logged in
+        pass
+    driver.implicitly_wait(1)
+    assert driver.find_element_by_xpath('//div[contains(text(), "testuser@gmail.com")]')
```
Great!

acrellin committed:
Use selenium-requests to wrap webdriver to enable API requests with authed driver

stefanv committed:
Print tables created

stefanv committed:
Run pytest via -m

stefanv committed:
Rename frontend_tests to test_frontend

stefanv committed:
Update peewee and psycopg2 dependencies

stefanv committed:
Correctly add test data

stefanv committed:
Only create each model once

stefanv committed:
Do not import User as AppUser as to avoid confusing peewee

stefanv committed:
Try importing models to help peewee determine creation order

stefanv committed:
Do not add fake test data

stefanv committed:
Use Postgres 9.6

stefanv committed:
Add back debug print of created tables

stefanv committed:
Refactor configuration to be explicit

stefanv committed:
Upgrade to Python 3.6

stefanv committed:
Ignore missing config files

stefanv committed:
Install correct cesium dependency; correctly load config for non-test runs

bnaul commented on tools/install_python_deps.py:
```diff
@@ -0,0 +1,25 @@
+with open(pjoin(os.path.dirname(__file__), '../requirements.txt')) as f:
+    deps = f.readlines()
+
+for dep in deps:
```
Combine as `for dep in open(pjoin(os.path.dirname(__file__), '../requirements.txt')):`?

stefanv committed:
Update ChromeDriver

stefanv committed:
Clean up install_python_deps with @bnaul's feedback

stefanv committed:
Slight cleanup to calling supervisorctl

bnaul committed:
Require cesium>=0.9.2

stefanv committed:
Now that we have a fixed cesium version, remove installer script

bnaul commented on cesium_app/config.py:
```diff
@@ -1,125 +1,51 @@
-    for key in ('paths', 'database', 'testing', 'docker'):
-        if key in cfg:
+        if ':' in key:
+            keys = key.split(':')
```
No need for the `if` block; `'a'.split(':') == ['a']`.

stefanv committed:
Slightly simplify configuration parser

acrellin committed:
Simpler dict.get() logic
